lib/galaxy_test/api/test_datasets.py

```diff
@@ -136,3 +136,16 @@ def test_tag_change(self):
         assert 'tag_c' in updated_hda['tags']
         assert 'name:tag_d' in updated_hda['tags']
         assert 'name:tag_e' in updated_hda['tags']
+
+    def test_update_datatype(self):
+        hda_id = self.dataset_populator.new_dataset(self.history_id)['id']
+        original_hda = self._get(
+            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=hda_id)).json()
+        assert original_hda['extension'] == 'txt'
+        update_payload = {
+            'extension': 'dat'
+        }
+        updated_hda = self._put(
+            "histories/{history_id}/contents/{hda_id}".format(history_id=self.history_id, hda_id=hda_id),
+            update_payload).json()
+        assert updated_hda['extension'] == 'dat'
```

Comment:
I think this shows it does not work, it doesn't update the metadata (probably just sets the extension in the database), and dat is not a valid extension, so this must fail.
For an implementation you might want to look at https://github.com/mvdbeek/galaxy/blob/cd7a2c3418d6540ab3bb3177b0b0216c54ebff56/lib/galaxy/webapps/galaxy/controllers/dataset.py#L365.

A test should then also demonstrate that datatype specific metadata has been set.

Reply:
Ops, that looked too easy! @simonbray You can probably move the implementation that Marius linked to a new `DatasetAssociationDeserializer.deserialize_extension()` .